### ISSUE TITLE
why tensorflow run slow in loop

### ISSUE THREAD

**Issue by MisayaZ** [tensorflow/tensorflow#1439]
I just run many times of sess.run in loop like this:

    input1 = tf.placeholder(tf.float32)
    input2 = tf.placeholder(tf.float32)
    sess = tf.Session()
    for image_num in range(128):
      image_result_table=np.zeros((96,32),dtype=np.float32)
      for sub_num in range(96):
          for centroids_num in range(32):
	       vector_a = local3_value[image_num,sub_num*4:(sub_num+1)*4]
	       vector_b = centroids[sub_num,centroids_num]
	       mat_result = tf.mul(input1, input2)
               reduce_sum = tf.reduce_sum(mat_result)
	       result =sess.run(reduce_sum,feed_dict={input1:vector_a, input2:vector_b})
               print(result)

at first ,it run fast. howerver, it become slower and slower. It takes almost an hour to get the whole result.

**Comment by girving** [comment]
For future issues, please use Markdown to format the code.  For this one, I edited it to indent by four spaces to preserve indentation.

**Comment by girving** [comment]
Your code snippet is incomplete, since you haven't shown us the definition of `local3_value` and `centroids`.  There are also some strange bits, such as allocating `image_result_table` but never using it.

**Comment by vrv** [comment]
@girving: the problem here is that the program is constantly adding more nodes to the graph because it is mixing graph construction with graph execution.

**Comment by girving** [comment]
Ah, right, forgot about our quadratic time behavior.  @MisayaZ: Move the `mat_result` and `reduce_sum` definitions outside the loops and you should be fine.

**Comment by MisayaZ** [comment]
Thanks

**Comment by hit-suit** [comment]
Hi @girving , when I run this code, the result shows that it just need 0.0273 second in the first epoch while 0.1528 second in the 200th epoch, and I don't know what is wrong with my code
`import tensorflow as tf
import numpy as np
import time

delta = np.random.uniform(size=(500, 500))
weight = tf.Variable(tf.truncated_normal([500, 500]))

x = tf.placeholder('float')
sess = tf.Session()
sess.run(tf.initialize_all_variables())

for epoch in range(200):
    t = time.time()
    for batch in range(10):
        update = weight.assign_add(x)
        sess.run(update, feed_dict={x: delta})
    print 'epoch %s done, time used %s' % (epoch, time.time()-t)`

Sorry for my bad english :D

**Comment by MisayaZ** [comment]
@hit-suit   "the program is constantly adding more nodes to the graph". I think this is the problem.

**Comment by bryanyzhu** [comment]
@MisayaZ @girving I have similar problem, the network is running slower and slower, and take up all the memory of my CPU, not GPU. So I was wondering if there is some convenient tool to check the graph, so I can know which nodes I am constantly adding to the graph? Thanks.

**Comment by mehdirezaie** [comment]
Hi Tensorflowers!

I am experiencing the same slow down issue by looping over sessions. In each loop I attempt to set a different seed for kernel initializer in `tf.layes.dense` , and I believe that "the program is constantly adding more nodes to the graph" causes this problem. In the following snipped I do my best to show how my program is structured. I set a global seed, and then generate a set of seeds in order to initialize my kernels differently in each run (or chain). The reason I am doing this is because I want to have independent trainings, and reproducible results. I was wondering if there is any way around to set the seed for the kernel initilizers without creating multiple nodes? At this point, I really appreciate any form of comments or help.

```
        global_seed = 1234
        np.random.seed(global_seed)
        seeds = np.random.randint(0, 4294967295, size=nchain)
        for ii in range(nchain): # loop over different runs/chains
            tf.set_random_seed(seeds[ii]) # set the seed
            # set up the model x [input] -> y [output] with two hidden layers
            x   = tf.placeholder(tf.float32, [None, nfeature])
            # nfeature, Units, scale are number of features, number of units on each hidden layer 
            # and regularization scale, respectively
            kernel_init0 = tf.random_normal_initializer(stddev=np.sqrt(1./(nfeature+1)), seed=seeds[ii])
            kernel_init1 = tf.random_normal_initializer(stddev=np.sqrt(2./(Units[0]+1)), seed=seeds[ii]) 
            kernel_init  = tf.random_normal_initializer(stddev=np.sqrt(2./(Units[1]+1)), seed=seeds[ii]) 
            y0 = tf.layers.dense(x,  units=Units[0], activation=tf.nn.relu, kernel_initializer=kernel_init0,
                                    kernel_regularizer= tf.contrib.layers.l2_regularizer(scale=scale))
            y1 = tf.layers.dense(y0, units=Units[1], activation=tf.nn.relu, kernel_initializer=kernel_init1,
                                    kernel_regularizer= tf.contrib.layers.l2_regularizer(scale=scale))
            y  = tf.layers.dense(y1, units=1,   activation=None,  kernel_initializer=kernel_init,
                                    kernel_regularizer= tf.contrib.layers.l2_regularizer(scale=scale))
            # placeholders for the input errorbar and label
            y_  = tf.placeholder(tf.float32, [None, 1])
            w   = tf.placeholder(tf.float32, [None, 1])

            #
            # objective function plus regularization term
            mse = tf.losses.mean_squared_error(y_, y, weights=w)
            l2_loss = tf.losses.get_regularization_loss()
            mse += l2_loss
            #
            global_step = tf.Variable(0, name='global_step', trainable=False)
            optimizer   = tf.train.AdamOptimizer(learning_rate)
            train_step  = optimizer.minimize(mse, global_step=global_step)            
            # 
            # initialize the NN
            sess = tf.InteractiveSession()
            tf.global_variables_initializer().run()            
            for epoch in range(number_of_epoch): # loop on training epochs
                # do training ...
            # evaluate something on test data    
            sess.close()
```
